File: 437PathSum3.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
var pathSum = function (root, sum) {
  var dfs = (node, sum, sum2) => {
    count = 0;
    if (!node) return;
    if (sum == sum2 + node.val) {
      count += 1;
      return count;
    }
    count += dfs(node.left, sum, sum2 + node.val);
    count += dfs(node.right, sum, sum2 + node.val);
    return count;
  };
  if (!root) return 0;

  return (
    pathSum(root.left, sum, 0) + dfs(root, sum, 0) + pathSum(root.right, sum, 0)
  );
};

File: 437PathSum3.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pathSum(self, root, sum):
        self.count = 0
        self.dfs(root, sum, 0)

    def dfs(self, node, sum, sum2):
        # if node is None:
        if sum == sum2:
            self.count += 1
        if node is None:
            return
        self.dfs(node.left, sum,  sum2+node.val)
        self.dfs(node.left, sum,  sum2)
        self.dfs(node.right, sum,  sum2+node.val)
        self.dfs(node.right, sum,  sum2)


a = Solution()
print(a.pathSum(root=[10, 5, -3, 3, 2, null, 11, 3, -2, null, 1], sum=8))
